### PR TITLE
feat(appsync): support S3 location for GraphQL schema

### PR DESCRIPTION
**Pull request by BryanPan342**
Add support for GraphQL Schema to define from an s3 location. For convenience, use the `appsync.Schema.fromS3Location` to specify the s3 location representing your schema.

<details>

<summary> Example </summary>

```ts
const bucket = new s3.Bucket(stack, 'Bucket');

const deploy = new s3deploy.BucketDeployment(stack, 'Deployment', {
  sources: [s3deploy.Source.asset(path.join(__dirname, 'schemas'))],
  destinationBucket: bucket,
});

const api = new appsync.GraphqlApi(stack, 'Api', {
  name: 'integ-test-s3',
  schema: appsync.Schema.fromS3Location({
    bucket,
    key: 'appsync.test.graphql',
  }),
});

api.node.addDependency(deploy);
```

</details>

**BREAKING CHANGES**: `appsync.Schema` changed to an abstract class
* **appsync**: No longer supporting `new appsync.Schema()`, instead use the static methods `appsync.Schema.fromCode()`, `appsync.Schema.fromAsset()`, `appsync.Schema.fromS3()`

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by MrArnoldPalmer** [comment]
@BryanPan342 we have had this one assigned to you for awhile. Are you still expecting changes or did we just forget to move it to myself or @shivlaks?

**Comment by BryanPan342** [comment]
@MrArnoldPalmer  honeslty i forgot but im pretty good with it as it is. 

i think we had some apprehensions with the underlying implications of assets.

File: packages/@aws-cdk/aws-appsync/README.md
Comment (skinny85) [inline packages/@aws-cdk/aws-appsync/README.md:340]: Can we show a real example here? For an actual file inside a Bucket?

File: packages/@aws-cdk/aws-appsync/lib/schema.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-appsync/lib/schema.ts:59]: I wonder if just a string is a good API here. We often represent S3 locations as pairs of `IBucket` and object key.

We also have the [`Location` interface in the S3 module](https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-s3.Location.html). Maybe that's a better abstraction than just a lonely string?

File: packages/@aws-cdk/aws-appsync/lib/schema.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-appsync/lib/schema.ts:81]: The fact that you have to write this validation suggests to me this API could be better.

Can we actually change this API here a little bit?

1. Make `Schema` abstract.
2. Create a subclass of `Schema`, `SchemaImpl`, that's module-private, and contains all of the logic here. It extends `Schema`, and takes `SchemaOptions` in its constructor argument. `SchemaOptions` should now also be module-private.
3. All of the static factory methods declare that they return `Schema` in their signature, but return `SchemaImpl` in their bodies.

This way, you will be in control of the invariants of `SchemaImpl`, and you won't have to write this defensive code. In fact, you can have a separate subclass of `Schema` for each of the static factory methods!

File: packages/@aws-cdk/aws-appsync/test/appsync-schema.test.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-appsync/test/appsync-schema.test.ts:36]: I would say the schema fails creation, "appsync fails" is kind of a weird statement 😃.

File: packages/@aws-cdk/aws-appsync/test/appsync-schema.test.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-appsync/test/appsync-schema.test.ts:298]: This is a weird test.. we have a separate `fromAsset()` static factory method for `Schema`, so why also use Asset here...?

File: packages/@aws-cdk/aws-appsync/test/appsync-schema.test.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-appsync/test/appsync-schema.test.ts:350]: These 2 tests are exactly the same, no?

File: packages/@aws-cdk/aws-appsync/test/appsync-schema.test.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-appsync/test/appsync-schema.test.ts:383]: These lines are repeated in like 5 tests in a row. Can you introduce a `describe()` block with a `beforeEach()` that gets rid of this duplication?

**Review by skinny85 — CHANGES_REQUESTED** [review]
Looks great! A couple of questions/suggestions around the API experience here, but in general super solid!

File: packages/@aws-cdk/aws-appsync/test/appsync-schema.test.ts
Comment (BryanPan342) [reply packages/@aws-cdk/aws-appsync/test/appsync-schema.test.ts:350]: An interface type is technically different than an object type but the validation is the same.. ill delete the interface one and rename the test to test against `IntermediateType`

File: packages/@aws-cdk/aws-appsync/lib/schema.ts
Comment (BryanPan342) [reply packages/@aws-cdk/aws-appsync/lib/schema.ts:81]: I'm all for this change, but it would be a breaking change because we might have users using the `Schema` api using `new Schema()`. Are we comfortable with introducing this breaking change?

File: packages/@aws-cdk/aws-appsync/lib/schema.ts
Comment (BryanPan342) [reply packages/@aws-cdk/aws-appsync/lib/schema.ts:59]: hm totally right.. ill use the IBucket and object key method alongside `s3UrlForObject`

File: packages/@aws-cdk/aws-appsync/test/appsync-schema.test.ts
Comment (skinny85) [reply packages/@aws-cdk/aws-appsync/test/appsync-schema.test.ts:350]: Let me ask this 🙂.

Is there anything that you can pass to `. addType()` there that would _not_ throw an exception?

File: packages/@aws-cdk/aws-appsync/lib/schema.ts
Comment (skinny85) [reply packages/@aws-cdk/aws-appsync/lib/schema.ts:81]: Yep, [`appsync` is still experimental](https://github.com/aws/aws-cdk/blob/1eb6de7913a27874a2c8c382900d6976fe7f0564/packages/%40aws-cdk/aws-appsync/package.json#L124-L125).

File: packages/@aws-cdk/aws-appsync/test/appsync-schema.test.ts
Comment (BryanPan342) [reply packages/@aws-cdk/aws-appsync/test/appsync-schema.test.ts:350]: ah yes that is very true

File: packages/@aws-cdk/aws-appsync/lib/schema.ts
Comment (BryanPan342) [reply packages/@aws-cdk/aws-appsync/lib/schema.ts:81]: Sweet, ill get it done today

**Comment by BryanPan342** [comment]
@skinny85 updated to reflect the new model

I did a reorg as well so sorry for the big PR but here is a quick summary:

* Moved `private.ts` to a `private/` directory to hold both MappingTemplates and Schema
* Removed dependency w/ `@aws-cdk/aws-s3-assets` and instead rely on `@aws-cdk/aws-s3` and `@aws-cdk/aws-s3-deployment`
* Moved all the testing `graphql` files under the `test/schema/` directory because I needed to upload to s3 and the `test/` dir was getting out of hand

I also did a manual integ test to make sure the schema is accessible by the GraphQL Api!

File: packages/@aws-cdk/aws-appsync/lib/schema.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-appsync/lib/schema.ts:25]: This is the wrong way around. `SchemaImpl` should extend `Schema`.

File: packages/@aws-cdk/aws-appsync/lib/schema.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-appsync/lib/schema.ts:42]: Make these positional arguments instead, the interface doesn't buy us much here.

File: packages/@aws-cdk/aws-appsync/lib/private/mapping-template.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-appsync/lib/private/mapping-template.ts:1]: Why is this file called `mapping-template.ts`? There's no such class/function in this file.

File: packages/@aws-cdk/aws-appsync/lib/private/schema-implementation.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-appsync/lib/private/schema-implementation.ts:117]: I believe you're doing this backwards.

All these methods should be `asbtract` in `Schema`, and then `SchemaImpl` should actually implement them.

File: packages/@aws-cdk/aws-appsync/lib/private/schema-implementation.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-appsync/lib/private/schema-implementation.ts:160]: Why the `Lazy` here?

File: packages/@aws-cdk/aws-appsync/lib/private/schema-implementation.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-appsync/lib/private/schema-implementation.ts:133]: Sorry, maybe I'm missing something.

Why does the `AssetSchema` read the file directly from disk? Doesn't this mean the Asset is not actually needed? (Assets are actually deployed to S3, but it seems like this is not needed here)

File: packages/@aws-cdk/aws-appsync/lib/private/schema.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-appsync/lib/private/schema.ts:1]: Can we call this file something like `schema-shape-addition.ts`? The current name really doesn't say anything.

File: packages/@aws-cdk/aws-appsync/lib/schema.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-appsync/lib/schema.ts:33]: I think you can cut the `Impl` from the name, and just call this `AssetSchema`.

File: packages/@aws-cdk/aws-appsync/lib/schema.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-appsync/lib/schema.ts:43]: Again, let's just call this `S3Schema`.

File: packages/@aws-cdk/aws-appsync/lib/schema.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-appsync/lib/schema.ts:53]: Again, let's just call this `CodeSchema`.

File: packages/@aws-cdk/aws-appsync/test/appsync-schema.test.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-appsync/test/appsync-schema.test.ts:317]: The descriptions of these tests could be improved. Remember that the parent descriptions are concatenated with the children. They should form proper sentences.

Something like:

```ts
  describe('API with Schema from S3', () => {
   // ...

  test('throws an exception when addQuery() is called on it', () => {
    // ...
```

File: packages/@aws-cdk/aws-appsync/test/integ.graphql-s3.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-appsync/test/integ.graphql-s3.ts:41]: Really don't like the fact that you have to call this yourself.

Should we have a `Schema` variant that takes in the S3 deployment type, that handles this for you?

**Review by skinny85 — CHANGES_REQUESTED** [review]
The PR is pretty tough to review, but I've seen worse 😉.

A few comments before this can be merged.

File: packages/@aws-cdk/aws-appsync/lib/private/schema-implementation.ts
Comment (BryanPan342) [reply packages/@aws-cdk/aws-appsync/lib/private/schema-implementation.ts:26]: ahhh i see thats what i was getting wrong... I think i misread your message hence the circular dependency stuff 🥲

File: packages/@aws-cdk/aws-appsync/lib/private/mapping-template.ts
Comment (BryanPan342) [reply packages/@aws-cdk/aws-appsync/lib/private/mapping-template.ts:1]: It holds helper functions for the mapping-templates!

Comment (BryanPan342) [reply packages/@aws-cdk/aws-appsync/lib/private/mapping-template.ts:1]: we could call it `mapping-templates-utils.ts`?

Comment (skinny85) [reply packages/@aws-cdk/aws-appsync/lib/private/mapping-template.ts:1]: How about something like `conditions.ts`, since those seem to be the classes present here?

File: packages/@aws-cdk/aws-appsync/lib/private/schema-implementation.ts
Comment (BryanPan342) [reply packages/@aws-cdk/aws-appsync/lib/private/schema-implementation.ts:133]: So originally the way AppSync was created was that we supported reading the `graphql` file. So in creating the static function that read a local file, we decided to follow the same scheme as Lambda `Code.fromAsset(...)`. Im guessing Asset is a misleading word?

File: packages/@aws-cdk/aws-appsync/test/integ.graphql-s3.ts
Comment (BryanPan342) [reply packages/@aws-cdk/aws-appsync/test/integ.graphql-s3.ts:41]: Hm.. i think this is residual code that I had before when I was trying to get the bucket token to resolve

File: packages/@aws-cdk/aws-appsync/lib/private/schema-implementation.ts
Comment (skinny85) [reply packages/@aws-cdk/aws-appsync/lib/private/schema-implementation.ts:133]: A little bit, in my opinion. "Asset" is a domain concept from the CDK (see [here](https://docs.aws.amazon.com/cdk/latest/guide/assets.html)).

I personally would have called that method `fromFile()`.

But let's maybe leave this one for now. It's kind of a separate issue. So I won't block the PR on it.

File: packages/@aws-cdk/aws-appsync/lib/private/mapping-template.ts
Comment (BryanPan342) [reply packages/@aws-cdk/aws-appsync/lib/private/mapping-template.ts:1]: Okay im good with that

File: packages/@aws-cdk/aws-appsync/test/integ.graphql-s3.ts
Comment (BryanPan342) [reply packages/@aws-cdk/aws-appsync/test/integ.graphql-s3.ts:41]: Actually nevermind, I retract that previous statement. Without the dependency between the deployment and schema, CDK doesn't know to that it needs to deploy the Schema after running the deployment. 

We currently have an API for the api called `addSchemaDependency` for the `GraphqlApi` construct.

If we want to make this a smoother process we could do the following:

```ts
export class S3Schema extends SchemaImpl {
  ...
  public constructor(bucket: IBucket, key: string, deployment?: BucketDeployment) {
     ...
     this.deployment = deployment.
  }

  public bind(api) {
    ...
    this.schema.node.addDependency(this.deployment);
  }
```

File: packages/@aws-cdk/aws-appsync/lib/private/schema-implementation.ts
Comment (BryanPan342) [reply packages/@aws-cdk/aws-appsync/lib/private/schema-implementation.ts:139]: Hm yeah okay I can live with that. If it makes your life easier, I can also split this PR into two to lower the file count!

File: packages/@aws-cdk/aws-appsync/test/integ.graphql-s3.ts
Comment (skinny85) [reply packages/@aws-cdk/aws-appsync/test/integ.graphql-s3.ts:38]: This needs to be a separate static factory option that takes in a `BucketDeployment` (not both an `IBucket` and a `BucketDeployment`, because `BucketDeployment` already has an `IBucket` defined).

Comment (BryanPan342) [reply packages/@aws-cdk/aws-appsync/test/integ.graphql-s3.ts:38]: ah yes that makes sense

Comment (BryanPan342) [reply packages/@aws-cdk/aws-appsync/test/integ.graphql-s3.ts:38]: Should that change occur in another PR? And keep this one focused on just `fromS3Location`? In fact, I wouldn't mind changing the static function name to `fromBucket`

Comment (skinny85) [reply packages/@aws-cdk/aws-appsync/test/integ.graphql-s3.ts:38]: Sure.

**Review by skinny85 — CHANGES_REQUESTED** [review]
Looks good, but I think I came up with a way to make `Schema` great 🙂. Let me know what you think!

File: packages/@aws-cdk/aws-appsync/lib/private/schema-implementation.ts
Comment (BryanPan342) [reply packages/@aws-cdk/aws-appsync/lib/private/schema-implementation.ts:55]: I dont want to create two schemas. So the conditional checks if it was created already.

File: packages/@aws-cdk/aws-appsync/lib/schema.ts
Comment (BryanPan342) [reply packages/@aws-cdk/aws-appsync/lib/schema.ts:50]: it existed before, so i kept it as to avoid breaking changes

File: packages/@aws-cdk/aws-appsync/lib/private/schema-implementation.ts
Comment (skinny85) [reply packages/@aws-cdk/aws-appsync/lib/private/schema-implementation.ts:55]: Why?

File: packages/@aws-cdk/aws-appsync/lib/private/schema-implementation.ts
Comment (BryanPan342) [reply packages/@aws-cdk/aws-appsync/lib/private/schema-implementation.ts:112]: wow.. i honestly did not know you could do this 😱  my mind is actually blown what

File: packages/@aws-cdk/aws-appsync/lib/schema.ts
Comment (skinny85) [reply packages/@aws-cdk/aws-appsync/lib/schema.ts:50]: Let's kill it.

File: packages/@aws-cdk/aws-appsync/lib/private/schema-implementation.ts
Comment (BryanPan342) [reply packages/@aws-cdk/aws-appsync/lib/private/schema-implementation.ts:112]: question, how would a user add this to the GraphqlApi? Because the Graphql API has the option of adding queries, and mutations to the schema as well!

https://github.com/aws/aws-cdk/blob/6b639bee925a9e552c405fde9d221d515e6023c3/packages/%40aws-cdk/aws-appsync/lib/graphqlapi.ts#L577-L643

If i set the type for schema to be a `Schema` it is immutable and thus cannot be targeted by the API. I guess what we **could** do is remove that feature entirely? Prevents new users from tripping and trying to add a Query if they are importing their schema locally.

File: packages/@aws-cdk/aws-appsync/lib/private/schema-implementation.ts
Comment (BryanPan342) [reply packages/@aws-cdk/aws-appsync/lib/private/schema-implementation.ts:55]: I mean.. if there is not point in it, why take up the memory to run it again? And I think the schema should be the same no? Unless there is some use case where you want to make multiple stacks that each have a schema that builds on top of itself?

File: packages/@aws-cdk/aws-appsync/lib/private/schema-implementation.ts
Comment (skinny85) [reply packages/@aws-cdk/aws-appsync/lib/private/schema-implementation.ts:112]: Maybe keep the API for now, but move the error code to `GraphqlApi` (basically, each of the methods does an `instanceof` check, and throws if the `schema` is not a `MutableSchema`)?

File: packages/@aws-cdk/aws-appsync/lib/private/schema-implementation.ts
Comment (skinny85) [reply packages/@aws-cdk/aws-appsync/lib/private/schema-implementation.ts:55]: Look at this code:

```ts
const schema = appsync.Schema.fromCode();
new GraphqlApi(this, 'Api1', { schema, });
new GraphqlApi(this, 'Api2', { schema, });
```

Because of  the caching that you do, the second `Api` does not have a Schema at all (its call to `Schema.bind()` returns the existing `CfnGraphQLSchema` with `apiId` filled with the ID of the first API).

I cant imagine this is the correct behavior.

**Comment by TheRealAmazonKendra** [comment]
This PR has been deemed to be abandoned, and will be closed. Please create a new PR for these changes if you think this decision has been made in error.

**Pull request closed without merging**
